soc/riscv/riscv-privileged/andes_v5/l2_cache.c

```diff
@@ -15,103 +15,93 @@
 #include <zephyr/arch/cpu.h>
 #include <zephyr/drivers/syscon.h>
 #include <zephyr/logging/log.h>
-LOG_MODULE_REGISTER(andes_v5_l2_cache, CONFIG_SOC_LOG_LEVEL);
-
-#if DT_NODE_EXISTS(DT_INST(0, andestech_l2c))
-
-/* SMU Configuration Register offset */
-#define SMU_SYSTEMCFG		0x08
+#include <zephyr/arch/riscv/csr.h>
 
-/* Register bitmask */
-#define SMU_SYSTEMCFG_L2C	BIT(8)
-
-/*
- * L2C Register Base Address
- */
+LOG_MODULE_REGISTER(andes_v5_l2_cache, CONFIG_SOC_LOG_LEVEL);
 
+/* L2C Register Base Address */
 #define ANDES_V5_L2C_BASE	DT_INST_REG_ADDR(0)
 
-/*
- * L2C Register Offset
- */
-
-#define L2C_CONFIG		0x00
-#define L2C_CTRL		0x08
+/* L2C Register Offset */
+#define L2C_CONFIG		(ANDES_V5_L2C_BASE + 0x00)
+#define L2C_CTRL		(ANDES_V5_L2C_BASE + 0x08)
 
-/*
- * L2C Helper Constant
- */
-
-/* L2 cache version */
-#define L2C_CONFIG_VERSION_SHIFT        24
-
-/* enable L2C */
+/* L2C Helper Constant */
+#define L2C_CONFIG_VER		GENMASK64(31, 24)
 #define L2C_CTRL_CEN		BIT(0)
 
-/* instruction prefetch depth */
+/* Instruction prefetch depth */
 #define IPFDPT_FIELD(x)		(x << 3)
 #define L2C_CTRL_IPFDPT_0	IPFDPT_FIELD(0)
 #define L2C_CTRL_IPFDPT_1	IPFDPT_FIELD(1)
 #define L2C_CTRL_IPFDPT_2	IPFDPT_FIELD(2)
 #define L2C_CTRL_IPFDPT_3	IPFDPT_FIELD(3)
 
-/* data prefetch depth */
+/* Data prefetch depth */
 #define DPFDPT_FIELD(x)		(x << 5)
 #define L2C_CTRL_DPFDPT_0	DPFDPT_FIELD(0)
 #define L2C_CTRL_DPFDPT_2	DPFDPT_FIELD(1)
 #define L2C_CTRL_DPFDPT_4	DPFDPT_FIELD(2)
 #define L2C_CTRL_DPFDPT_8	DPFDPT_FIELD(3)
 
+#if DT_HAS_COMPAT_STATUS_OKAY(andestech_atcsmu100)
+/* SMU Register offset */
+#define SMU_SYSTEMCFG		0x08
+
+/* SMU Helper Constant */
+#define SMU_SYSTEMCFG_L2C	BIT(8)
+#endif
+
 static void andes_v5_l2c_enable(void)
 {
-	unsigned long mcache_ctl;
-	volatile uint64_t *l2c_ctrl =
-		INT_TO_POINTER(ANDES_V5_L2C_BASE + L2C_CTRL);
+	uint32_t l2c_ctrl = sys_read32(L2C_CTRL);
 
-	__asm__ volatile ("csrr %0, %1"
-		: "=r" (mcache_ctl) : "i" (NDS_MCACHE_CTL));
+	/* Enable L2C if I-cache or D-cache is enabled */
+	if (csr_read(NDS_MCACHE_CTL) & BIT_MASK(2)) {
+		uint32_t l2c_config = sys_read32(L2C_CONFIG);
 
-	/* Enable L2 cache if L1 I/D cache enabled */
-	if (mcache_ctl & (BIT(1) | BIT(0))) {
-		volatile uint64_t *l2c_config =
-			INT_TO_POINTER(ANDES_V5_L2C_BASE + L2C_CONFIG);
+		/* Memory barrier, flush all I/D-Cache before setting L2C */
+		__asm__ volatile ("fence.i");
```

Comment:
Please use the newly introduced barrier API for this. See `include/zephyr/sys/barrier.h`

Reply:
Andes core needs to execute fence.i to flush all I/D cache before updating the L2 cache function.
The barrier API in RISC-V seems only support fence instruction.
Maybe I can modify the comment to avoid confusion.

Reply:
Right but why you cannot use `fence` here instead of `fence.i`? AFAIK `fence` should be just a bit more restrictive.

Reply:
In Andes core,  `fence` instruction is used to order device I/O and memory accesses as viewed by other RISC-V harts and external devices, but it does not perform any operation on the local I/D cache.
`fence.i` is used to synchronizing the L1 and L2 cache due to the hardware limitation of Andes L2 cache.